emscripten/webarkitJsfeat.cpp

```diff
@@ -87,6 +89,138 @@ emscripten::val load_jpeg_data(std::string filename) {
+        filename);
+  }
+  if (strcmp(ext, "jpeg") == 0 || strcmp(ext, "jpg") == 0 ||
+      strcmp(ext, "jpe") == 0) {
+    webarkitLOGi("Waiting for the jpeg...");
+    webarkitLOGi("Reading JPEG file...");
+    ar2UtilDivideExt(filename, buf1, buf2);
+    jpegImage = ar2ReadJpegImage(buf1, buf2);
+    if (jpegImage == NULL) {
+      webarkitLOGe(
+          "Error: unable to read JPEG image from file '%s'. Exiting.\n",
+          filename);
+    }
+    webarkitLOGi("   Done.");
+
+    if (jpegImage->nc != 1 && jpegImage->nc != 3) {
+      ARLOGe(
+          "Error: Input JPEG image is in neither RGB nor grayscale format. "
+          "%d bytes/pixel %sformat is unsupported. Exiting.\n",
+          jpegImage->nc, (jpegImage->nc == 4 ? "(possibly CMYK) " : ""));
+    }
+    webarkitLOGi("JPEG image number of channels: '%d'", jpegImage->nc);
+    webarkitLOGi("JPEG image width is: '%d'", jpegImage->xsize);
+    webarkitLOGi("JPEG image height is: '%d'", jpegImage->ysize);
+    webarkitLOGi("JPEG image, dpi is: '%d'", jpegImage->dpi);
+
+    if (jpegImage->dpi == 0.0f) {
+      webarkitLOGw(
+          "JPEG image '%s' does not contain embedded resolution data, and no "
+          "resolution specified on command-line.",
+          filename);
+    }
+
+  } else if (strcmp(ext, "png") == 0) {
+    webarkitLOGe(
+        "Error: file has extension '%s', which is not supported for "
+        "reading. Exiting.\n",
+        ext);
+    free(ext);
+  }
+  webarkitLOGi("Image done!");
+
+  JSLOGi("Starting detection routine...");
+
+  Orb orb;
+  Imgproc imgproc;
+  detectors::Detectors detectors;
+  auto width = jpegImage->xsize;
+  auto height = jpegImage->ysize;
+  std::unique_ptr<Matrix_t> lev0_img = std::make_unique<Matrix_t>(width, height, ComboTypes::U8C1_t);
+  std::unique_ptr<Matrix_t> lev_img = std::make_unique<Matrix_t>(width, height, ComboTypes::U8C1_t);
+  Array<std::unique_ptr<Matrix_t>> pattern_corners;
+
+  auto sc0 = std::min(max_pattern_size / height, max_pattern_size / width);
+  //new_width = (jpegImage->ysize * sc0) | 0;
+  //new_height = (jpegImage->xsize * sc0) | 0;
+  auto num_train_levels = 4;
+  
+  JSLOGi("Converting the RGB image to GRAY...");
+  
+  imgproc.grayscale_internal<u_char, Matrix_t>(jpegImage->image, width, height, lev0_img.get(), Colors::COLOR_RGB2GRAY);
+  
+  JSLOGi("Image converted to GRAY.");
+
+  Array<KeyPoints*> lev_corners(num_train_levels);
+  //Array<std::unique_ptr<KeyPoints>> lev_corners;
+  //Array<std::unique_ptr<Matrix_t>> pattern_descriptors;
+
+  for (lev = 0; lev < num_train_levels; ++lev) {
+    // what we should do with this code?
+    // pattern_corners[lev] = [];
+    // lev_corners = pattern_corners[lev];
+
+    // preallocate corners array
+    //i = (new_width * new_height) >> lev;
+    i = (width * height) >> lev;
+    std::cout << i << std::endl;
+    while (--i >= 0) {
+      lev_corners[lev]->set_size(i);
+      //lev_corners[lev]->allocate();
```

Comment:
no need to allocate if we create the  vector: Array<KeyPoints*> lev_corners(num_train_levels);